### PR TITLE
Configurable Per-World Void Damage Height Offset & Amount

### PR DESCRIPTION
**Pull request by Axionize**

### CODE REVIEW

File: CONTRIBUTING.md
Comment (lynxplay) [inline CONTRIBUTING.md]: What happened here?

File: patches/api/0490-voidDamageThreshold-Height-at-which-entiies-take-voi.patch
Comment (lynxplay) [inline patches/api/0490-voidDamageThreshold-Height-at-which-entiies-take-voi.patch]: The method isn't returning anything, `@return` does not make sense here.
It is also missing a `@param voidDamageAmount` description.

Same for the threshold.

File: patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch
Comment (lynxplay) [inline patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch]: Why was this changed?
Entity#level() exists already as a getter.

File: CONTRIBUTING.md
Comment (RealTriassic) [inline CONTRIBUTING.md]: Looks like you accidentally edited this?

File: patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch
Comment (lynxplay) [inline patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch]: In general, this breaks compatibility with worlds that do not conform to the expected "size" used when you calculated the default of -128.

This could be fixed by simply using DoubleOr.Default, allowing the default to evaluate to the specific min - 64 value.

File: patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch
Comment (lynxplay) [inline patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch]: These values seem completely unused, the logic in entity uses the world config, no the CB worlds values.

File: patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch
Comment (lynxplay) [inline patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch]: This value seems completely unused.

**Review by lynxplay — CHANGES_REQUESTED** [review]
Welcome to paper :tada: 
Thank you for your first PR.

Left a couple of comments that need addressing before something like this can be merged.

File: patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch
Comment (Machine-Maker) [reply patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch]: I might prefer just allowing the *offset* from the min build height to be configured. So the default would then be -64. Changing it to 0 would then make damage happen right at the min build height.

**Comment by Axionize** [comment]
I've fixed all the issues you've shown me and changed the PR to be an offset from minBuildHeight() as you said you wanted.

**Comment by lynxplay** [comment]
Do you have a specific use-case for adding API getters/setters?
We'd rather hold off on adding more methods until we have a strong API specific usecase, given this seems rather admin specific with the config values.

The API should already be able to deal void damage via the damage cause API if needed and these methods add to the general clutter that is the World interface already.

**Comment by lynxplay** [comment]
Beyond that, we can combine enabled and damage into a FloatOr.Disabled (needs to be created)

**Comment by Axionize** [comment]
Yes, dynamically loading/building worlds with WorldCreator from region files on disk, mainly for minigames. 

Currently you can generate a custom world at runtime with a custom generator, biomeProvider, etc... but obviously void kill height can not be set when creating the world, much less at runtime.

An alternative may be to modify WorldCreator.java to support passing in a custom world.yaml config settings (and potentially reloading those settings at runtime in World.java), but this is obviously much simpler and I would very much enjoy not having to take on such a task.

**Comment by lynxplay** [comment]
Eh sounds fair, I guess.
I'll chat it over with the rest of the team.

You can deal void damage perfectly fine via https://jd.papermc.io/paper/1.21.1/org/bukkit/entity/Damageable.html#damage(double,org.bukkit.damage.DamageSource) and https://jd.papermc.io/paper/1.21.1/org/bukkit/damage/DamageType.html#OUT_OF_WORLD respectively tho.

Maybe it is worth consolidating the void damage configuration into a single interface type, e.g.

World#getVoidDamageConfiguration()
World#setVoidDamageConfiguration(WorldVoidDamageConfiguration)

interface WorldVoidDamageConfiguration {
  
  boolean isVoidDamageEnabled();
  double getVoidDamagePerTick();
  float getMinWorldHeightOffset();

  static WorldVoidDamageConfiguration vanilla() { }
  static WorldVoidDamageConfiguration enabled(double damage, float minWorldHeightOffset) { }
  static WorldVoidDamageConfiguration disabled() { }
}

but I'll get back to concrete next steps once we have more team input on this.
For now, could you please address machines comment regarding the "threshold".
It cannot be a constant y-value, it should be expressed as an offset from the min build height, as the default world configuration otherwise breaks custom height worlds.

**Comment by Axionize** [comment]
I've made it an offset from the build height as requested, defaulting to -64 blocks from the minBuildHeight.

**Review by lynxplay — APPROVED** [review]

**Review by Machine-Maker — APPROVED** [review]

**Pull request merged**
